### DISCUSSION TITLE
Updating the Value source

### DISCUSSION THREAD

**Discussion by emmalcg** [react-querybuilder/react-querybuilder#545] [Q&A] [answered]
Hi, I am building a pretty complex query to export to jsonLogic and have all custom components over-riding the defaults to meet our use cases. 
I am wondering if there is a way to update the valueSource of a rule when the ValueEditor value changes if my validation check passes. (Basically running a check to see if the value inputted is a valid data point from our api). As if it is a datapoint entered within the valueEditor - I need this to have the value source of 'Field' so that the value is exported within the var object of jsonLogic. 

I am also over-riding the Field Selector component with a text-input / data point selection modal, and i am wondering if I provide the `Fields` props to the `QueryBuilder` (so that I have access to the data type) is there a way so that my field isn't being pre-populated within the first value on this Fields array? (Setting adding a new rule to have all blank values)

Currently right now I have manipulated the Query so that both `field` and `value` are objects `field: { value: , dataType: }` `value: { value: , dataType: }` instead of strings and have written my own jsonLogic rule processor to check if the `dataType` is not undefined on the `value` to render it in a rule object- and this is working (with many ignore ts comments 😓 ), but now running into issues when the operator is `between` or `notBetween` and i need to be rendering two of these value inputs that would store two value objects on the query `value` and then transform them into two var objects within jsonLogic.

Sorry for the long note - any suggestions are appreciated!!!

**Comment by jakeboone02** [comment, marked as answer]
> Hi, I am building a pretty complex query to export to jsonLogic and have all custom components over-riding the defaults to meet our use cases. I am wondering if there is a way to update the valueSource of a rule when the ValueEditor value changes if my validation check passes. (Basically running a check to see if the value inputted is a valid data point from our api). As if it is a datapoint entered within the valueEditor - I need this to have the value source of 'Field' so that the value is exported within the var object of jsonLogic.

Setting one property from the control of another property, as you're trying to do (setting the `valueSource` from the `value` editor), is not straightforward (yet! the next version should make this much easier). The best way I've found is to pass the full `query` object down through the [`context` prop](https://react-querybuilder.js.org/docs/components/querybuilder#context), then run `setQuery` to update it in full (not just the `value` property like `handleOnChange` would do) using the [`update` method](https://react-querybuilder.js.org/docs/utils/misc#update).

Here's an example that sets `valueSource` to `"field"` as soon as the `value` matches a field name other than the current field:

https://stackblitz.com/edit/github-m2rcdw?file=src%2FApp.tsx

> I am also over-riding the Field Selector component with a text-input / data point selection modal, and i am wondering if I provide the `Fields` props to the `QueryBuilder` (so that I have access to the data type) is there a way so that my field isn't being pre-populated within the first value on this Fields array? (Setting adding a new rule to have all blank values)

Sounds like `autoSelectField={false}` is what you want. [Docs here](https://react-querybuilder.js.org/docs/components/querybuilder#autoselectfield).

> Currently right now I have manipulated the Query so that both `field` and `value` are objects `field: { value: , dataType: }` `value: { value: , dataType: }` instead of strings and have written my own jsonLogic rule processor to check if the `dataType` is not undefined on the `value` to render it in a rule object- and this is working (with many ignore ts comments 😓 ), but now running into issues when the operator is `between` or `notBetween` and i need to be rendering two of these value inputs that would store two value objects on the query `value` and then transform them into two var objects within jsonLogic.

I would strongly recommend against using an object for `field` properties on rules, and (slightly less strongly) recommend against it for `value` if you can avoid it. Think of the `field` property (and `operator` and `combinator` for that matter) as an ID reference, not an information store.

If you need to get properties like `dataType` from the `Field` definition, try the (soon-to-be documented) `getOption` method:

```ts
getOption(fields, 'field_name_here')?.dataType;
```

(That's pretty much the same as `fields.find(f => f.name === 'field_name_here')` if your `fields` array is flat, but `getOption` will also handle arrays of option groups.)

> Sorry for the long note - any suggestions are appreciated!!!

No worries. Hope I could help!

**Reply by emmalcg** [reply]
thanks so much! these answers are helpful :) 

One more follow up, for `autoSelectField={false}` , is there a way to turn off this feature? 

> When the empty field option is selected, the operator selector and value components will not be rendered for that rule.

I need these fields to always show up 🌞

**Reply by jakeboone02** [reply]
> is there a way to turn off this feature?

All that feature really does is add an "empty" option to the front of the `fields` array, so you could replicate it by adding your own "empty" option without hiding the operator selector and value editor. You just need to make sure the `name` of the option doesn't match `translations.fields.placeholderName`, which is `"~"` by default. That's what triggers the operator selector and value editor to be hidden. (The value editor is hidden when `operator` matches `translations.operators.placeholderName`. The corresponding prop is [`autoSelectOperator`](https://react-querybuilder.js.org/docs/components/querybuilder#autoselectoperator).)

Something like this should work:

```ts
const fields: Field[] = [
  { name: '~~~', label: '------' },
  ...otherFields,
];
```

Then just take `field: '~~~'` to mean "not selected".

**Reply by emmalcg** [reply]
thanks @jakeboone02 that makes sense!!

**Reply by emmalcg** [reply]
@jakeboone02 sorry 😬  Wondering why this type error would be happening for generating the fields for valueSources ? 

`Type '{ name: string; label: string; dataType: DataType | undefined; valueSources: string[]; }' is not assignable to type 'Field<string, string, string, Option<string>, Option<string>>'.
  Types of property 'valueSources' are incompatible.
    Type 'string[]' is not assignable to type 'ValueSources | ((operator: string) => ValueSources) | undefined'.`

```
...filteredDataPoints.map((point) => ({
        name: point.id || '',
        label: point.id || '',
        dataType: point.dataType,
        valueSources: ['value', 'field'],
      })),
```

**Reply by jakeboone02** [reply]
Try this:

```ts
...filteredDataPoints.map((point) => ({
        name: point.id || '',
        label: point.id || '',
        dataType: point.dataType,
        valueSources: ['value', 'field'] as const,
      })),
```

The problem is `['value', 'field'] ` gets "widened" to `string[]`. I think you could also do this:

```ts
...filteredDataPoints.map((point) => ({
        name: point.id || '',
        label: point.id || '',
        dataType: point.dataType,
        valueSources: ['value', 'field'],
      } as Field)),
```

**Reply by emmalcg** [reply]
thanks @jakeboone02 ! the `as Field` worked perfectly :)
